psl/psl-core/src/validate/validation_pipeline/validations/models.rs

```diff
@@ -274,30 +274,84 @@ pub(super) fn id_client_name_does_not_clash_with_field(model: ModelWalker<'_>, c
     }
 }
 
-pub(super) fn schema_attribute(model: ModelWalker<'_>, ctx: &mut Context<'_>) {
-    match (model.schema(), ctx.datasource) {
-        (Some((_, span)), _) if !ctx.connector.has_capability(ConnectorCapability::MultiSchema) => ctx.push_error(
-            DatamodelError::new_static("@@schema is not supported on the current datasource provider", span),
-        ),
-        (Some((schema_name, span)), Some(ds)) if !ds.has_schema(schema_name) => {
+pub(super) fn schema_is_defined_in_the_datasource(model: ModelWalker<'_>, ctx: &mut Context<'_>) {
+    if !ctx.preview_features.contains(crate::PreviewFeature::MultiSchema) {
+        return;
+    }
+
+    if !ctx.connector.has_capability(ConnectorCapability::MultiSchema) {
+        return;
+    }
+
+    let datasource = match ctx.datasource {
+        Some(ds) => ds,
+        None => return,
+    };
+
+    let (schema_name, span) = match model.schema() {
+        Some(tuple) => tuple,
+        None => return,
+    };
+
+    if datasource.has_schema(schema_name) {
+        return;
+    }
+
     ctx.push_error(DatamodelError::new_static(
         "This schema is not defined in the datasource. Read more on `@@schema` at https://pris.ly/d/multi-schema",
         span,
     ))
-        },
-        (None, _)
-            if ctx
-                .db
-                .schema_flags()
-                .contains(parser_database::SchemaFlags::UsesSchemaAttribute) =>
-        {
-            ctx.push_error(DatamodelError::new_static(
-                "This model is missing an `@@schema` attribute.",
-                model.ast_model().span(),
-            ))
-        }
-        _ => (),
+}
+
+pub(super) fn schema_attribute_supported_in_connector(model: ModelWalker<'_>, ctx: &mut Context<'_>) {
+    if !ctx.preview_features.contains(crate::PreviewFeature::MultiSchema) {
+        return;
+    }
+
+    if ctx.connector.has_capability(ConnectorCapability::MultiSchema) {
+        return;
+    }
+
+    let (_, span) = match model.schema() {
+        Some(tuple) => tuple,
+        None => return,
+    };
+
+    ctx.push_error(DatamodelError::new_static(
+        "@@schema is not supported on the current datasource provider",
+        span,
+    ));
+}
+
+pub(super) fn schema_attribute_missing(model: ModelWalker<'_>, ctx: &mut Context<'_>) {
+    if !ctx.preview_features.contains(crate::PreviewFeature::MultiSchema) {
+        return;
     }
+
+    if !ctx.connector.has_capability(ConnectorCapability::MultiSchema) {
+        return;
+    }
+
+    if !ctx
+        .db
+        .schema_flags()
+        .contains(parser_database::SchemaFlags::UsesSchemaAttribute)
+    {
+        return;
+    }
+
+    if ctx.connector.is_provider("mysql") {
+        return;
+    }
```

Comment:
I don't understand this

Reply:
moved from here https://github.com/prisma/prisma-engines/pull/3264/files#diff-68eb7688702988516bd4df93dc1255a3cab8fcb0103d3eea48eb9a4ff9756623L22